File: java/socketd/src/main/java/org/noear/socketd/transport/core/internal/Frames.java
package org.noear.socketd.transport.core.internal;

import org.noear.socketd.SocketD;
import org.noear.socketd.transport.core.EntityMetas;
import org.noear.socketd.transport.core.Flags;
import org.noear.socketd.transport.core.Frame;
import org.noear.socketd.transport.core.Message;
import org.noear.socketd.transport.core.entity.EntityDefault;
import org.noear.socketd.transport.core.entity.StringEntity;

/**
 * 帧工厂
 *
 * @author noear
 * @since 2.0
 * */
public class Frames {
    /**
     * 构建连接帧
     *
     * @param url 连接地址
     */
    public static final Frame connectFrame(String sid, String url) {
        EntityDefault entity = new EntityDefault();
        //添加框架版本号
        entity.meta(EntityMetas.META_SOCKETD_VERSION, SocketD.version());
        return new Frame(Flags.Connect, new MessageDefault().sid(sid).event(url).entity(entity));
    }

    /**
     * 构建连接确认帧
     *
     * @param connectMessage 连接消息
     */
    public static final Frame connackFrame(Message connectMessage) {
        EntityDefault entity = new EntityDefault();
        //添加框架版本号
        entity.meta(EntityMetas.META_SOCKETD_VERSION, SocketD.version());
        return new Frame(Flags.Connack, new MessageDefault().sid(connectMessage.sid()).event(connectMessage.event()).entity(entity));
    }

    /**
     * 构建 ping 帧
     */
    public static final Frame pingFrame() {
        return new Frame(Flags.Ping, null);
    }

    /**
     * 构建 pong 帧
     */
    public static final Frame pongFrame() {
        return new Frame(Flags.Pong, null);
    }

    /**
     * 构建关闭帧（一般用不到）
     */
    public static final Frame closeFrame() {
        return new Frame(Flags.Close, null);
    }

    /**
     * 构建告警帧（一般用不到）
     */
    public static final Frame alarmFrame(Message from, String alarm) {
        MessageDefault message = new MessageDefault();

        if (from != null) {
            //如果有来源消息，则回传元信息
            message.sid(from.sid());
            message.event(from.event());
            message.entity(new StringEntity(alarm).metaString(from.metaString()));
        } else {
            message.entity(new StringEntity(alarm));
        }

        return new Frame(Flags.Alarm, message);
    }
}

File: python/socketd/socketd/transport/core/Frames.py
from socketd import SocketD
from socketd.transport.core.Entity import Entity
from socketd.transport.core.Flags import Flags
from socketd.transport.core.EntityMetas import EntityMetas
from socketd.transport.core.HandshakeDefault import HandshakeInternal
from socketd.transport.core.Message import Message
from socketd.transport.core.entity.EntityDefault import EntityDefault
from socketd.transport.core.Frame import Frame
from socketd.transport.core.entity.MessageBuilder import MessageBuilder
from socketd.transport.core.entity.StringEntity import StringEntity


class Frames:

    @staticmethod
    def connect_frame(sid: str, url: str, metaMap: dict[str, str]) -> Frame:
        entity = StringEntity(url)
        # 添加框架版本号
        entity.meta_map_put(metaMap)
        entity.meta_put(EntityMetas.META_SOCKETD_VERSION, SocketD.version())

        message = MessageBuilder().sid(sid).event(url).entity(entity).build()

        return Frame(Flags.Connect, message)

    @staticmethod
    def connack_frame(handshake: HandshakeInternal) -> Frame:
        entity = EntityDefault()
        # 添加框架版本号
        entity.meta_map_put(handshake.get_out_meta_map())
        entity.meta_put(EntityMetas.META_SOCKETD_VERSION, SocketD.version())
        entity.data_set(handshake.get_source().entity().data())

        message = (MessageBuilder().sid(handshake.get_source().sid())
                   .event(handshake.get_source().event())
                   .entity(entity).build())

        return Frame(Flags.Connack, message)

    @staticmethod
    def ping_frame() -> Frame:
        return Frame(Flags.Ping, None)

    @staticmethod
    def pong_frame() -> Frame:
        return Frame(Flags.Pong, None)

    @staticmethod
    def close_frame(_code: int) -> Frame:
        messageBuilder = MessageBuilder()
        messageBuilder.entity(EntityDefault().meta_put("code", str(_code)))

        return Frame(Flags.Close, messageBuilder.build())

    @staticmethod
    def alarm_frame(_from: Message, alarm: Entity) -> Frame:
        messageBuilder = MessageBuilder()

        if _from:
            entity = EntityDefault()
            entity.meta_string_set(_from.meta_string())
            entity.data_set(alarm.data());
            entity.meta_map_put(alarm.meta_map())

            messageBuilder.sid(_from.sid());
            messageBuilder.event(_from.event());
            messageBuilder.entity(entity)
        else:
            messageBuilder.entity(alarm)

        return Frame(Flags.Alarm, messageBuilder.build())

    @staticmethod
    def pressure_frame(_from: Message, pressure: Entity) -> Frame:
        messageBuilder = MessageBuilder()

        if _from:
            entity = EntityDefault()
            entity.meta_string_set(_from.meta_string())
            entity.data_set(pressure.data());
            entity.meta_map_put(pressure.meta_map())

            messageBuilder.sid(_from.sid());
            messageBuilder.event(_from.event());
            messageBuilder.entity(entity)
        else:
            messageBuilder.entity(pressure)

        return Frame(Flags.Pressure, messageBuilder.build())
